### ISSUE TITLE
How to finetune only one class?

### ISSUE THREAD

**Issue by Stephenfang51** [ultralytics/yolov5#3913]
## ❔Question
Thanks for your excellent work in advanced !
what if I want to finetune only one class of my own dataset?
if I have[0,1,2,3] classes in my dataset, I want to add more data for "0" class, and do not want to train other my whole data again
and also I wish the model output 4 classes. Any advice ?

Thanks so much !

**Comment by glenn-jocher** [comment]
@Stephenfang51 just update your dataset as you see fit and retrain on it.

**Comment by kelectro** [comment]
I am not sure what you mean @glenn-jocher. Just update the specific class and retrain using the whole dataset?

**Comment by glenn-jocher** [comment]
@kelectro what's your question exactly?

**Comment by kelectro** [comment]
I have a trained model in 4 classes and works fine in 3/4.  I would like to know if its possible to tune only one specific class instead of  retraining the model on the whole dataset.

**Comment by glenn-jocher** [comment]
@kelectro 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. If you have a poorly performing class the easiest method to improve it is to collect more data for it.

[Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/](https://github.com/ultralytics/yolov5/issues/2844#issuecomment-851338384)

**Comment by kelectro** [comment]
Thank you for the very usefull guide!

**Comment by TonyTran86** [comment]
> I have a trained model in 4 classes and works fine in 3/4. I would like to know if its possible to tune only one specific class instead of retraining the model on the whole dataset.

As I understand in case the other classes not labeled when they existed in the image of dataset then the result will read as "negative cases".

**Comment by visionKinger** [comment]
> ## ❔Question
> Thanks for your excellent work in advanced ! what if I want to finetune only one class of my own dataset? if I have[0,1,2,3] classes in my dataset, I want to add more data for "0" class, and do not want to train other my whole data again and also I wish the model output 4 classes. Any advice ?
> 
> Thanks so much !
> 
> ## Additional context

In my opinion,  you can freeze the layers but the classification layer. In the classification layers, you should use Tensor.data.requiresgrad=False, in case not to learn other classes. And after calculating the gradient, you should clear the grad

**Comment by glenn-jocher** [comment]
@visionKinger hi there,

To fine-tune only one class of your own dataset without retraining the entire model, you can follow these steps:

1. Freeze the layers of the pre-trained model except for the classification layer for the desired class.

2. Set the `requires_grad` property of the parameters in the frozen layers to False. This ensures that the gradients are not computed for those layers during the backward pass, preventing them from being updated during training.

3. During the forward pass, calculate the gradient only for the classification layer of the desired class. You can set `requires_grad` to True for the classification layer of the desired class, allowing it to be updated during training.

4. After calculating the gradient, clear the gradients for the frozen layers. This prevents them from being accumulated and interfering with the optimization process.

By following these steps, you can focus on fine-tuning only the desired class without retraining the entire model, and the model will continue to output predictions for all four classes.

I hope this advice helps! Let me know if you have any further questions.

Best regards,

**Comment by ggloss** [comment]
> @visionKinger hi there,
> 
> To fine-tune only one class of your own dataset without retraining the entire model, you can follow these steps:
> 
> 1. Freeze the layers of the pre-trained model except for the classification layer for the desired class.
> 2. Set the `requires_grad` property of the parameters in the frozen layers to False. This ensures that the gradients are not computed for those layers during the backward pass, preventing them from being updated during training.
> 3. During the forward pass, calculate the gradient only for the classification layer of the desired class. You can set `requires_grad` to True for the classification layer of the desired class, allowing it to be updated during training.
> 4. After calculating the gradient, clear the gradients for the frozen layers. This prevents them from being accumulated and interfering with the optimization process.
> 
> By following these steps, you can focus on fine-tuning only the desired class without retraining the entire model, and the model will continue to output predictions for all four classes.
> 
> I hope this advice helps! Let me know if you have any further questions.
> 
> Best regards,

Hi @glenn-jocher, could you please recommend a guide with code to follow these steps? I've trained a YOLO model with the train.py for 10 classes and I just want to fine-tune the model for 2 classes on 10 of these classes .

**Comment by glenn-jocher** [comment]
Hi @ggloss,

Fine-tuning specific classes in YOLOv5 is not a standard feature, and the architecture does not support freezing individual classification layers for each class as it uses a single output layer for all classes. However, you can approach this by training on a dataset that includes only the new images for the class you wish to improve, while keeping the other classes in the dataset to prevent catastrophic forgetting.

Here's a general approach:

1. Prepare a new dataset where the class you want to fine-tune is well-represented, and include a smaller number of examples from the other classes.
2. Load your pre-trained weights with the `--weights` flag.
3. Train the model on the new dataset using the `train.py` script. The model will update based on the new data, which should improve performance on the class of interest while maintaining knowledge of the other classes.

Remember to monitor the performance on a validation set that represents all classes to ensure that the model does not forget the other classes.

For specific code guidance, please refer to the Ultralytics documentation and examples, as providing code is beyond the scope of this platform.

**Comment by ggloss** [comment]
@glenn-jocher many thanks for your advice! I'll try to reproduce the presented approach.

**Comment by glenn-jocher** [comment]
You're welcome, @ggloss! I'm glad to hear you found the advice helpful. If you encounter any issues or have further questions as you proceed, feel free to reach out. Best of luck with your fine-tuning efforts! Happy training! 🚀😊
